services/app-api/src/postgres/contractAndRates/updateDraftContractWithRates.ts

```diff
@@ -0,0 +1,375 @@
+    }
+}
+
+// Update the given draft
+// * can change the set of draftRates
+// * set the formData
+async function updateDraftContractWithRates(
+    client: PrismaClient,
+    args: UpdateContractArgsType
+): Promise<ContractType | NotFoundError | Error> {
+    const { contractID, formData, rateFormDatas } = args
+    const {
+        submissionType,
+        submissionDescription,
+        programIDs,
+        populationCovered,
+        riskBasedContract,
+        stateContacts,
+        supportingDocuments,
+        contractType,
+        contractExecutionStatus,
+        contractDocuments,
+        contractDateStart,
+        contractDateEnd,
+        managedCareEntities,
+        federalAuthorities,
+        modifiedBenefitsProvided,
+        modifiedGeoAreaServed,
+        modifiedMedicaidBeneficiaries,
+        modifiedRiskSharingStrategy,
+        modifiedIncentiveArrangements,
+        modifiedWitholdAgreements,
+        modifiedStateDirectedPayments,
+        modifiedPassThroughPayments,
+        modifiedPaymentsForMentalDiseaseInstitutions,
+        modifiedMedicalLossRatioStandards,
+        modifiedOtherFinancialPaymentIncentive,
+        modifiedEnrollmentProcess,
+        modifiedGrevienceAndAppeal,
+        modifiedNetworkAdequacyStandards,
+        modifiedLengthOfContract,
+        modifiedNonRiskPaymentArrangements,
+        inLieuServicesAndSettings,
+    } = formData
+
+    try {
+        return await client.$transaction(async (tx) => {
+            // Given all the Contracts associated with this draft, find the most recent submitted
+            const currentRev = await client.contractRevisionTable.findFirst({
+                where: {
+                    contractID: contractID,
+                    submitInfoID: null,
+                },
+                include: {
+                    contract: true,
+                    draftRates: {
+                        include: includeDraftRates,
+                    },
+                },
+            })
+            if (!currentRev) {
+                const err = `PRISMA ERROR: Cannot find the current rev to update with contract id: ${contractID}`
+                console.error(err)
+                return new NotFoundError(err)
+            }
+
+            const stateCode = currentRev.contract.stateCode as StateCodeType
+            const ratesFromDB = currentRev.draftRates.map((rate) =>
+                rateRevisionToDomainModel(rate.revisions[0])
+            )
+            const updateRates =
+                rateFormDatas && sortRatesForUpdate(ratesFromDB, rateFormDatas)
+
+            if (updateRates) {
+                for (const rateRevision of updateRates.upsertRates) {
+                    // Check if the rate exists
+                    // - We don't know if the rate revision exists in the DB we just know it's not connected to the contract.
+                    // - toProtoBuffer gives every rate revision a UUID if there isn't one, so we cannot rely on revision id.
```

Comment:
🤔 Could we just move where ids are generated on new rate infos? I get that `toProtobuffer` currently has a workaround added for our packages with shared rates feature but not sure we need that hack in the location its currently placed at.  For example, could we move ID generation for new rates farther down in the code, closer to the db layer - maybe to the `updateHealthPlanRevision` postgres handler instead? That way it would  also live alongside the legacy code it was built for. It would be feature flagged and contained there. It wouldn't be carried into new world where I thought we have postgres rate revision IDs we trust and those are the rate ids

Reply:
I also think changing `toProtoBuffer` not to generate UUIDs and instead having the resolver generate it doesn't give us much gain. Generating the UUIDs in`toProtoBuffer` was done to give each rate in the protobuffer an ID we could identify it. So to me, this was strictly for how we can identify rates in a protobuf.

Now, moving to Postgres, we have relationships, and manually generating the UUID is not necessary. Rates without a UUID then would mean it's a new rate. But since we live in `ProtoGres` world we just make it work for now by checking if the rate exists and seems to be less work.